File: Difficulty: Medium/Unique Paths in a Grid/unique-paths-in-a-grid.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
//User function Template for C++

class Solution {
  public:
  #define mod (int)(1e9+7)
     int solve(int i,int j,int n,int m,vector<vector<int>> &grid,vector<vector<int>> &dp){
         if(i>=n  || j>=m || grid[i][j]==0)return 0;
         if(i==n-1 and j==m-1)return 1;
         if(dp[i][j]!=-1)return dp[i][j];
         int d=solve(i+1,j,n,m,grid,dp);
            int r=solve(i,j+1,n,m,grid,dp);
            return dp[i][j]=(d+r)%mod;
     }
    int uniquePaths(int n, int m, vector<vector<int>> &grid) {
        // code here
        vector<vector<int>>dp(n+1,vector<int>(m+1,-1));
        return solve(0,0,n,m,grid,dp);
        
    }
};

//{ Driver Code Starts.

int main() {
    int t;
    cin >> t;
    while (t--) {
        int n,m,x;
        cin>>n>>m;
        
        vector<vector<int>> grid(n,vector<int>(m));
        
        for(int i=0; i<n; i++){
            for(int j=0; j<m; j++){
                cin>>grid[i][j];
            }
        }

        Solution ob;
        cout << ob.uniquePaths(n,m,grid) << endl;
    }
    return 0;
}
// } Driver Code Ends

File: Difficulty: Medium/Unique Paths in a Grid/unique-paths-in-a-grid.py
class Solution:
    def uniquePaths(self, grid):
        if grid[0][0]==1:
            return 0
        hth=len(grid)
        wth=len(grid[0])
        from functools import lru_cache
        @lru_cache(None)
        def dp(x=wth-1,y=hth-1):
            nonlocal grid
            if x==0 and y==0:
                return 1
            ret=0
            if y-1>=0 and grid[y-1][x]==0:
                ret+=dp(x,y-1)
            if x-1>=0 and grid[y][x-1]==0:
                ret+=dp(x-1,y)
            return ret
        return dp()